libraries/chain/fork_database.cpp

```diff
@@ -277,20 +277,22 @@ namespace eosio::chain {
          EOS_RETHROW_EXCEPTIONS( fork_database_exception, "serialized fork database is incompatible with configured protocol features" )
       }
 
+      const auto prev_indx = index.template get<by_best_branch>().begin();
```

Comment:
You don't need the variable `prev_indx` or the call to `index.template get<by_best_branch>().begin();` here.

Reply:
The `.begin()` should be almost no cost. Do you think the `n == head_impl(include_root_t::no)` is actually better? Seems like the `by_best_branch` is pretty clear.

Reply:
Why do we need this line at all?

Reply:
For the ` prev_indx != index.template get<by_best_branch>().begin()` below.

Reply:
But we don't need `prev_indx` below.

If we inserted a new `bsp`, and that `bsp` is the new best head, then we do have a new best head.
The last line can be:

```
      return inserted.second && head_impl(include_root_t::no) == n;
```

Reply:
Right. My point is:
```cpp
const auto prev_indx = index.template get<by_best_branch>().begin();
...
return inserted.second && prev_indx != index.template get<by_best_branch>().begin();
```
Is more efficient when inserted which is the normal case. But I can change to `head_impl(include_root_t::no) == n` if you think it is better.

Reply:
That is not a lookup it just returns the named index reference. It is implemented as `return *this`.

Reply:
Oh I see you're right! Thanks for enlightening me :-).
I think I still prefer my version (or  `return inserted.second && n == *index.template_ get<by_best_branch>().begin();`) because it doesn't rely on the specific behavior of the multi-index implementation. It looks like all the indices are updated whenever an element is added, but I though it could be done lazily maybe (only when the indices are accessed), although now that I think more about it it probably would be tricky to implement.

Reply:
Also relies on stable iterators which `multi_index` also provides. 

Since you think it is clearer, I'll go with comparing it to `n`.

If comparing with `n`, I think `head_impl(include_root_t::no) == n` should be almost as efficient as accessing the index directly (one additional if check).